### PR TITLE
ListNotifier Tests - notificationMode types

### PR DESCRIPTION
**Pull request by KhaderMurtaja**
### Test Results ⬇️ 

list_notifier_test_results

### CODE REVIEW

File: lib/src/list_notifier.dart [hunk 1]
```diff
@@ -5,12 +5,23 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
 import 'package:functional_listener/functional_listener.dart';
 
 /// A List that behaves like `ValueNotifier` if its data changes.
-/// it does not compare the elements on bulk operations
-/// If you set [notifyIfEqual] to `false` it will compare if a value passed value
-///  passed is equal to the existing value.
-/// like `list[5]=4` if the content at index 4 is equal to 4 and only call
+/// It does not compare the elements on bulk operations
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: Actually this isn't longer the case. for most bulk operations it will compare elements.

Comment (escamoteur) [reply lib/src/list_notifier.dart]: The only exceptions are 
grafik
perhaps we should list this functionnames there.  Most probably these functions will always change the List. 
I could not move the comments before the method name because then the inherited API doc of List won't appear when you hover over a method

File: lib/src/list_notifier.dart [hunk 2]
```diff
@@ -5,12 +5,23 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
...
+///
+/// If you set [notificationMode] to [normal] `ListNotifier` will compare if a
+/// value passed is equal to the existing value.
+/// Like `list[5]=4` if the content at index 4 is equal to 4 and only call
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: Like in `list[5]=4` it tests if the content at index 4 is equal to 4 and only call

**Review by escamoteur — CHANGES_REQUESTED** [review]
Thanks for your work, just some minor changes, then we are done

File: lib/src/list_notifier.dart [hunk 3]
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
...
 /// A List that behaves like `ValueNotifier` if its data changes.
-/// It does not compare the elements on bulk operations
+/// It does not compare the elements on bulk operations like `addAll` or
+/// `replaceRange` but will notify listeners if the list changes.
 ///
-/// If you set [notificationMode] to [normal] `ListNotifier` will compare if a
-/// value passed is equal to the existing value.
-/// Like `list[5]=4` if the content at index 4 is equal to 4 and only call
-/// `notifyListeners` if they are not equal.
+/// If you want to compare the elements you can use the
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: why shortening this? The original comment might be improved but explained it better

File: lib/src/list_notifier.dart [hunk 4]
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
...
+/// `CustomNotifierMode.normal` mode and provide a custom equality function.
 ///
-/// If you set [notificationMode] to [manual] you have to call `notifyListeners`
-/// manually to notify all listeners.
-/// Like `list.notifyListeners();` if you want to notify all listeners.
+/// If you want to prevent the list from notifying listeners you can use the
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: same, improve the original but don't reduce information

File: lib/src/list_notifier.dart [hunk 5]
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
...
+/// `CustomNotifierMode.manual` mode.
 ///
-/// If you set [notificationMode] to [always] `ListNotifier` will always notify
-/// all listeners if the list changes.
-/// Like `list[5]=4` will always notify all listeners.
+/// If you want to notify listeners on every change you can use the
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: same

File: lib/src/list_notifier.dart [hunk 6]
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
...
+/// `CustomNotifierMode.always` mode.
 ///
-/// To allow atomic changes `ListNotifier` supports a single level of
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: why remove this?

File: lib/src/list_notifier.dart [hunk 7]
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
...
-/// transactions
+/// The functions that will always notify listeners unless [notificationMode]
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: I would add `will always notifiy listeners without comparing the existing value`

File: lib/src/list_notifier.dart [hunk 8]
```diff
@@ -32,10 +33,10 @@ class ListNotifier<T> extends DelegatingList<T>
   ///  passed is equal to the existing value.
   /// like `list[5]=4` if the content at index 4 is equal to 4 and only call
   /// `notifyListeners` if they are not equal. To prevent users from wondering
-  /// why their UI doesn't update if they haven't overritten the equality operator
-  /// the default is [always].
-  /// [customEquality] can be used to set your own criteria for comparing when choosing
-  /// [normal] as [notifierMode].
+  /// why their UI doesn't update if they haven't overwritten the equality
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: actually the correct way would be `overridden` if we talk about overriding an method.

File: test/list_notifier_test.dart [hunk 9]
```diff
@@ -5,8 +5,8 @@ import '../lib/src/list_notifier.dart';
 
 void main() {
   group("Tests for the ListNotifier's methods", () {
-    late ListNotifier list;
-    List result = [];
+    late ListNotifier<int> list;
```
Comment (escamoteur) [inline test/list_notifier_test.dart:7]: yep, good to use types here

Comment (escamoteur) [reply test/list_notifier_test.dart:7]: As the new default it to notify always we should set mode to normal for the tests.

**Review by escamoteur — CHANGES_REQUESTED** [review]

File: lib/src/list_notifier.dart (hunk 4 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: have done that.

File: lib/src/list_notifier.dart (hunk 6 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: It was mistakenly, sorry!

File: lib/src/list_notifier.dart (hunk 7 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: Done ✅

File: lib/src/list_notifier.dart (hunk 8 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: Fixed

File: lib/src/list_notifier.dart [hunk 10]
```diff
@@ -4,13 +4,30 @@ import 'package:collection/collection.dart';
...
-/// If you set [notifyIfEqual] to `false` it will compare if a value passed value
-///  passed is equal to the existing value.
-/// like `list[5]=4` if the content at index 4 is equal to 4 and only call
-/// `notifyListeners` if they are not equal.
-/// To allow atomic changes `ListNotifier` supports a single level of transactions
+/// `ListNotifier` is a List that behaves similarly to `ValueNotifier` when its
+/// data changes.
+/// It does not compare the elements during bulk operations.
+///
+/// If [notifyIfEqual] is set to `false`, `ListNotifier` will compare the new
```
Comment (escamoteur) [inline lib/src/list_notifier.dart]: why do I see again `notifyIfEqual` in the comments, that does no longer exist. We now have the `notificationMode` so correct would be if [notificationMode] is set to [normal].

The default is [always] so that the UI will be updated on every write access of the List

File: test/list_notifier_test.dart (hunk 9 above)
Comment (escamoteur) [reply test/list_notifier_test.dart:7]: you didn't address this comment

File: lib/src/list_notifier.dart (hunk 5 above)
Comment (escamoteur) [reply lib/src/list_notifier.dart]: include the example again

File: lib/src/list_notifier.dart (hunk 7 above)
Comment (escamoteur) [reply lib/src/list_notifier.dart]: " without comparing the existing value" still  missing

**Review by escamoteur — CHANGES_REQUESTED** [review]
still some open points. sorry for nitpicking.
Amal didn't push anything yet.

File: test/list_notifier_test.dart (hunk 9 above)
Comment (KhaderMurtaja) [reply test/list_notifier_test.dart:7]: @escamoteur, what do you mean by "address this comment"? 🤔👀

File: lib/src/list_notifier.dart (hunk 7 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: @escamoteur, it's already there at lines `[35, 35]` 🙋🏼‍♂️ 

screenshot

File: lib/src/list_notifier.dart (hunk 5 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: @escamoteur, it's already there at lines `[19, 22]` 🙋🏼‍♂️ 

screenshot

File: lib/src/list_notifier.dart (hunk 10 above)
Comment (KhaderMurtaja) [reply lib/src/list_notifier.dart]: @escamoteur, fixed it 🙋🏼‍♂️

**Comment by KhaderMurtaja** [comment]
> still some open points. sorry for nitpicking. Amal didn't push anything yet.

@escamoteur, never mind I enjoy enhancing the documentation 😁, it has to be perfect for better understanding 💯
I have addressed your comments, I hope you find it ready to merge now unless you have more points 🙏🏼

File: lib/src/list_notifier.dart (hunk 5 above)
Comment (escamoteur) [reply lib/src/list_notifier.dart]: It's not bad if we repeat some information

File: test/list_notifier_test.dart (hunk 9 above)
Comment (escamoteur) [reply test/list_notifier_test.dart:7]: I meant this comment:
>As the new default it to notify always we should set mode to normal for the tests.

if you change that you will have to review the tests again

**Review by escamoteur — CHANGES_REQUESTED** [review]
Docs are fine now, but the tests should use notificationtype.normal which means they have to be tested again then

**Comment by KhaderMurtaja** [comment]
> Docs are fine now, but the tests should use notificationtype.normal which means they have to be tested again then

@escamoteur, All tests are covered now, and even more tests have been added such as: `"customEquality handles null values correctly"`.

Check out the [PR description](https://github.com/escamoteur/listenable_collections/pull/7#issue-2317626275) for the Test Results ✅.

**Review by escamoteur — CHANGES_REQUESTED** [review]
I fear you misunderstood me.
As most test here were written with the old default of NotificationMode.normal, you should change the Notification mode of the ListNotifier that is used through all of this tests in setup to NotificationMode.Normal
Before you change any of the tests
